### PR TITLE
Update9.1.0

### PR DESCRIPTION
**Pull request by alugowski**
Update to https://github.com/GraphBLAS/python-suitesparse-graphblas/pull/127

### CODE REVIEW

**Comment by alugowski** [comment]
I got working builds. Briefly:

* Besides the `suitesparse/` include prefix, new GraphBLAS now installs to different locations than it used to. The new default location on Unix systems is `/usr/local`. On mac 14+ the install now requires `sudo`, and of course `sudo` does not exist on Windows. Must handle both. Also, Conda still installs to the old locations, so must still support old `sys.prefix` install location in addition to `/usr/local`. I tried explicitly setting the install location, but that caused issues with wheel repair that I couldn't figure out.
* I had to disable the JIT because it is hopelessly broken. On Windows it fails on the very first build line. On other systems it causes segfaults during tests. Let's mess with the JIT separately. @DrTimothyAldenDavis is this expected?
* macOS: I cleaned up the removal of the `libomp` multi-arch hack. GitHub Actions `macos-14` and later runners are Apple Silicon, so I cleaned up the build process to make use of this. Now wheels are built and tested natively on both x86 (old macos-12 runner) and ARM (macos-14 runner). Note that now the earliest supported macOS versions are 12 (x86) and 14 (ARM).
* Windows: Need to pass special flags to the compiler to tell `GraphBLAS.h` to use MSVC-style complex values (`"-DGxB_HAVE_COMPLEX_MSVC"`). @DrTimothyAldenDavis please don't make your users do this. It's a simple one-liner to detect MSVC and set the appropriate value [here](https://github.com/DrTimothyAldenDavis/GraphBLAS/blob/9bdf19dea3de59210979a9691807cfabccdf04ea/Include/GraphBLAS.h#L167), I beg you to automate stuff like this and not dump it on your users. Default usage results in errors like this
```
    C:/GraphBLAS\include\suitesparse\GraphBLAS.h(166): error C2061: syntax error: identifier 'GxB_FC32_t'
    C:/GraphBLAS\include\suitesparse\GraphBLAS.h(166): error C2059: syntax error: ';'
    C:/GraphBLAS\include\suitesparse\GraphBLAS.h(167): error C2061: syntax error: identifier 'GxB_FC64_t'
    C:/GraphBLAS\include\suitesparse\GraphBLAS.h(167): error C2059: syntax error: ';'
    C:/GraphBLAS\include\suitesparse\GraphBLAS.h(1663): error C2146: syntax error: missing ')' before identifier 'identity'
    C:/GraphBLAS\include\suitesparse\GraphBLAS.h(1663): error C2081: 'GxB_FC32_t': name in formal parameter list illegal
    C:/GraphBLAS\include\suitesparse\GraphBLAS.h(1663): error C2061: syntax error: identifier 'identity'
    C:/GraphBLAS\include\suitesparse\GraphBLAS.h(1663): error C2059: syntax error: ';'
    C:/GraphBLAS\include\suitesparse\GraphBLAS.h(1664): error C2059: syntax error: ')'
```

**Comment by DrTimothyAldenDavis** [comment]
The jit should work in all platforms. It passes my ci tests in the main suitesparse github. 

You shouldn't have to tell Windows to use its MSVC complex types. That is supposed to be configured by cmake automatically. There's a cmake module that checks that, in the GraphBLAS/cmake_folder/*.cmake if I recall. (On my phone at the moment)

**Comment by alugowski** [comment]
The JIT is a tangential issue to the new version build, but just for documentation here are the errors that forced me to disable the JIT. I don't see the [GraphBLAS CI setup](https://github.com/DrTimothyAldenDavis/GraphBLAS/actions) (also no workflows in `.github/` dir), so I can't compare or peak at the upstream CI to help identify what the setup difference is.

My guess is a dependency on some system tooling (compiler?) that isn't met on the intentionally bare build runners. Someone more familiar with how the JIT works should take a crack at it.

[Windows build error](https://github.com/GraphBLAS/python-suitesparse-graphblas/actions/runs/9343575516/job/25713378326#step:9:351):
```
   + make -j4
  [  0%] Building C object JITpackage/CMakeFiles/grb_jitpackage.dir/Source/grb_jitpackage.c.obj
  [  1%] Linking C executable grb_jitpackage.exe
  [  1%] Built target grb_jitpackage
  [  1%] Generating compressed sources for JIT compiler...
  make[2]: *** [JITpackage/CMakeFiles/GB_JITpackage.dir/build.make:308: D:/a/python-suitesparse-graphblas/python-suitesparse-graphblas/GraphBLAS-9.1.0/JITpackage/GB_JITpackage.c] Error 127
  make[2]: *** Deleting file 'D:/a/python-suitesparse-graphblas/python-suitesparse-graphblas/GraphBLAS-9.1.0/JITpackage/GB_JITpackage.c'
  make[1]: *** [CMakeFiles/Makefile2:153: JITpackage/CMakeFiles/GB_JITpackage.dir/all] Error 2
  make: *** [Makefile:136: all] Error 2
```

[macOS test crash](https://github.com/GraphBLAS/python-suitesparse-graphblas/actions/runs/9343613161/job/25713478170#step:9:4306):
```
   ======================= 1 passed, 8 deselected in 0.30s ========================
  ============================= test session starts ==============================
  platform darwin -- Python 3.8.10, pytest-8.2.1, pluggy-1.5.0 -- /private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/bin/python
  cachedir: .pytest_cache
  rootdir: /private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/test_cwd
  collecting ... collected 9 items
  
  tests/test_doctest.py::test_run_doctests PASSED                          [ 11%]
  tests/test_exceptions.py::test_check_status PASSED                       [ 22%]
  tests/test_io.py::test_serialize_matrix Fatal Python error: Aborted
  
  Current thread 0x0000000114894600 (most recent call first):
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/suitesparse_graphblas/tests/test_io.py", line 104 in test_serialize_matrix
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/_pytest/python.py", line 162 in pytest_pyfunc_call
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/pluggy/_callers.py", line 103 in _multicall
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/pluggy/_manager.py", line 120 in _hookexec
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/pluggy/_hooks.py", line 513 in __call__
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/_pytest/python.py", line 1632 in runtest
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/_pytest/runner.py", line 173 in pytest_runtest_call
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/pluggy/_callers.py", line 103 in _multicall
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/pluggy/_manager.py", line 120 in _hookexec
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/pluggy/_hooks.py", line 513 in __call__
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/_pytest/runner.py", line 241 in <lambda>
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/_pytest/runner.py", line 341 in from_call
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/_pytest/runner.py", line 240 in call_and_report
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/_pytest/runner.py", line 135 in runtestprotocol
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/_pytest/runner.py", line 116 in pytest_runtest_protocol
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/pluggy/_callers.py", line 103 in _multicall
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/pluggy/_manager.py", line 120 in _hookexec
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/pluggy/_hooks.py", line 513 in __call__
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/_pytest/main.py", line 364 in pytest_runtestloop
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/pluggy/_callers.py", line 103 in _multicall
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/pluggy/_manager.py", line 120 in _hookexec
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/pluggy/_hooks.py", line 513 in __call__
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/_pytest/main.py", line 339 in _main
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/_pytest/main.py", line 285 in wrap_session
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/_pytest/main.py", line 332 in pytest_cmdline_main
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/pluggy/_callers.py", line 103 in _multicall
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/pluggy/_manager.py", line 120 in _hookexec
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/pluggy/_hooks.py", line 513 in __call__
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/_pytest/config/__init__.py", line 178 in main
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/lib/python3.8/site-packages/_pytest/config/__init__.py", line 206 in console_main
    File "/private/var/folders/dg/ntz3m8yx7nz375nbbkd59r3c0000gn/T/cibw-run-gd20c66h/cp38-macosx_x86_64/venv-test-x86_64/bin/pytest", line 8 in <module>
  /bin/sh: line 1: 20049 Abort trap: 6           pytest -v --pyargs suitesparse_graphblas
```

**Comment by alugowski** [comment]
> You shouldn't have to tell Windows to use its MSVC complex types. That is supposed to be configured by cmake automatically. There's a cmake module that checks that, in the GraphBLAS/cmake_folder/*.cmake if I recall.

Unclear why this didn't work for us.

However this should not be set at GraphBLAS built stage, because the header is being installed systemwide and the user's compiler isn't known yet. One could imagine building GraphBLAS with MSVC then later building with MinGW. Which version should a package manager ship?

**Comment by DrTimothyAldenDavis** [comment]
The GraphBLAS CI is not in the GraphBLAS repo.  It's in SuiteSparse, which includes GraphBLAS, LAGraph, and all my other packages:  https://github.com/DrTimothyAldenDavis/SuiteSparse ,
and both Windows and Mac work fine with the JIT in that CI.

I have no idea why the MSVC complex type was not found properly.  It is detected by this file during cmake build time:
https://github.com/DrTimothyAldenDavis/GraphBLAS/blob/stable/cmake_modules/GraphBLAS_complex.cmake

The Windows error you're seeing has nothing to do with the JIT itself as used inside GraphBLAS.  It is the small program that runs at build time to create the JITpackage.c source file, before the GraphBLAS library is compiled.  It runs fine on the SuiteSparse CI on windows, with many compilers:
https://github.com/DrTimothyAldenDavis/SuiteSparse/actions/runs/9333486242/job/25690563509#step:16:2054
which reports the following on Windows with the MS cl compiler:

```
1995/5428] Building C object GraphBLAS\JITpackage\CMakeFiles\grb_jitpackage.dir\Release\Source\grb_jitpackage.c.obj
[1996/5428] Linking C executable GraphBLAS\JITpackage\Release\grb_jitpackage.exe
[1997/5428] Generating compressed sources for JIT compiler...
Processing 242 input files ...
Total uncompressed: 1909807 bytes
Total compressed:   330640 bytes
Compression:        0.173127
```

I don't understand the Mac segfaults.  That looks like strictly a python thing.  I don't see any GraphBLAS errors there.  I wouldn't know how to fix those.

In the typical case that the same compiler is used to build GraphBLAS is also used by the end user, on Windows, then all should be automatic.  If the user needs a different complex type, then they could use -DGxB_COMPLEX_H and then use their own typedefs, I suppose.  I could see about making that easier to use one compiler (say clang) to build GraphBLAS and another (say cl) for the end user application, or visa versa.  This is all an unfortunate workaround for a single broken compiler (MSVC cl), which doesn't support ANSI C99 complex types.  All other compilers on Windows do support them (clang, gcc, icx, and so on).

**Comment by DrTimothyAldenDavis** [comment]
Here's the reason why we added a feature test at build time to determine which complex types are supported: https://github.com/DrTimothyAldenDavis/SuiteSparse/pull/754

See the discussion around that PR.

I think the main issue is that the compiled graphblas.dll library would have one complex type, and the user application should use the same.  If the end user wants a different complex type, then they need another graphblas.dll compiled with the appropriate complex type.

Otherwise things are fragile.

**Comment by alugowski** [comment]
Ah yes, I remember writing the hacks to make MSVC complex types work for the Windows wheels :) I believe there are some deeper guarantees between compatibility of complex types cross-language and likely cross-compiler (like it's guaranteed safe to typecast C and C++ complex types because they're explicitly speced to have the same memory layout), but I suppose the fragility will stay. I appreciate the cmake approach, but beware that it appears to break sometimes.

**Comment by alugowski** [comment]
> The Windows error you're seeing has nothing to do with the JIT itself as used inside GraphBLAS. It is the small program that runs at build time to create the JITpackage.c source file, before the GraphBLAS library is compiled. It runs fine on the SuiteSparse CI on windows, with many compilers: https://github.com/DrTimothyAldenDavis/SuiteSparse/actions/runs/9333486242/job/25690563509#step:16:2054 which reports the following on Windows with the MS cl compiler:

Not sure what to say, whatever it is breaks the build. Error 127 likely means a file is missing. The build environment that `cibuildwheel` provides is intentionally minimal, so maybe there is some unspecified dependency?

> I don't understand the Mac segfaults. That looks like strictly a python thing. I don't see any GraphBLAS errors there. I wouldn't know how to fix those.

It's not a Python error. It's a SIGABORT on a call into GraphBLAS: https://github.com/GraphBLAS/python-suitesparse-graphblas/blob/045d489732638ce416c007bfdad102ce61d7de53/suitesparse_graphblas/tests/test_io.py#L104

Somebody will have to turn on logging, or replicate it locally.

Note that Python wheels are expected to be self-contained and need to bundle all libraries that they use. The "repair" process I mentioned above involves scouring all referenced dynamic libraries, copying them into the wheel, and relocating their references so they load properly. Dynamic calls into the system will break things, as the user's system will be different than the one that built it, and things like compilers can't be assumed to exist. `cibuildwheel` does tests properly; in a clean environment not contaminated by things installed by the build step. This way if tests pass there it's very likely the wheel will work on users' systems. To me this error smells like some hidden dependency is missing at runtime.

I suggest keeping the JIT off here until a brave soul volunteers to go through the pain.

**Comment by alugowski** [comment]
@eriknw Here's the options for the JIT that I can think of.

The JIT needs to be configured with a compiler, which is done at GraphBLAS build time by [GraphBLAS_JIT_configure.cmake](https://github.com/DrTimothyAldenDavis/GraphBLAS/blob/9bdf19dea3de59210979a9691807cfabccdf04ea/cmake_modules/GraphBLAS_JIT_configure.cmake). Of course that's a non-starter for Python because the user's machine is different than the build machine, but AFAIK it's possible to set a new config at runtime.

One option is to `pip install cmake` and run that configure script at import time (or an explicit `enable_jit()`), then set that config. As long as there's safe a fallback, explicit documentation, and a guided experience through possible failures then this can work. Shoaib Kamil's SEJITS did this a decade ago and we used it to compile user-defined semirings.

Another is to tie in to the Python compiler packages like numba (ships LLVM and tackles the versioning and compatibility nighmare), or pythran or even Cython. This would be a nicer user experience, but I haven't dug too much into how much of GraphBLAS the JIT compiles. This may work if it's a user callback or even a kernel, but likely not if it's a large chunk.

Another is to only support the JIT on conda. I believe they have some sort of compiler guarantees in their environments, so maybe then you _can_ assume the build-time compiler is available at runtime. May require some magic to handle different absolute paths to conda environments, but I'm sure they have some best practices for this.

Apart from perhaps conda, JIT support isn't going to be a simple flip-a-flag affair. Either way, I recommend handling it separately, I opened https://github.com/GraphBLAS/python-suitesparse-graphblas/issues/129 to track it.

I'll let you decide if what we have here is enough or if more work needs to be done.

**Comment by DrTimothyAldenDavis** [comment]
Yes, there is a way to set the compiler used in the JIT at run time, with GrB_set.  I have many options for controlling the JIT with GrB_set / GrB_get:

```
    GxB_JIT_C_COMPILER_NAME = 7024,  // CPU JIT C compiler name
    GxB_JIT_C_COMPILER_FLAGS = 7025, // CPU JIT C compiler flags
    GxB_JIT_C_LINKER_FLAGS = 7026,   // CPU JIT C linker flags
    GxB_JIT_C_LIBRARIES = 7027,      // CPU JIT C libraries
    GxB_JIT_C_PREFACE = 7028,        // CPU JIT C preface
    GxB_JIT_C_CONTROL = 7029,        // CPU JIT C control
    GxB_JIT_CACHE_PATH = 7030,       // CPU/CUDA JIT path for compiled kernels
    GxB_JIT_C_CMAKE_LIBS = 7031,     // CPU JIT C libraries when using cmake
    GxB_JIT_USE_CMAKE = 7032,        // CPU JIT: use cmake or direct compile
    GxB_JIT_ERROR_LOG = 7033,        // CPU JIT: error log file
```

The defaults for all of these are found by cmake when GraphBLAS itself is compiled and so by default I use the compiler, flags, and so on, that were used to build GraphBLAS (seems like a reasonable default, but of course it is different for the python-graphblas wrapper).  I configure the GraphBLAS/Config/GB_config.h.in file to create GraphBLAS/Config/GB_config.h.  When GrB_init starts, it sets all those strings to the defaults in GB_config.h, and if desired they can be changed later at any time, in the end user application.  I only use the definitions in GB_config.h at GrB_init time.  So none of those strings in GB_config.h are baked in stone; all of them can be changed later at run time.

So that means at runtime, the python wrapper can make any decisions it likes.  It can try to set up its own defaults when python starts. It could also accept some option from the python user, like `gb.compiler('/usr/bin/mynewcompiler')` or something, as you like.

**Comment by DrTimothyAldenDavis** [comment]
> It's not a Python error. It's a SIGABORT on a call into GraphBLAS:
> 
> [python-suitesparse-graphblas/suitesparse_graphblas/tests/test_io.py](https://github.com/GraphBLAS/python-suitesparse-graphblas/blob/045d489732638ce416c007bfdad102ce61d7de53/suitesparse_graphblas/tests/test_io.py#L104)
> 
> Line 104 in [045d489](https://github.com/GraphBLAS/python-suitesparse-graphblas/commit/045d489732638ce416c007bfdad102ce61d7de53)
> 
>  lib.GrB_Matrix_eWiseAdd_BinaryOp(C[0], NULL, NULL, _eq_ops[T], A[0], B[0], NULL), 
> Somebody will have to turn on logging, or replicate it locally.

The burble can help.  I could also try to replicate it on my system, if I had some help with getting things installed and compiled (I rarely use python so I'd need some help).

**Comment by DrTimothyAldenDavis** [comment]
I guess my last two comments should be added to #129.  I'll copy them there.

**Comment by DrTimothyAldenDavis** [comment]
Regarding the complex type in GraphBLAS.h:


> Ah yes, I remember writing the hacks to make MSVC complex types work for the Windows wheels :) I believe there are some deeper guarantees between compatibility of complex types cross-language and likely cross-compiler (like it's guaranteed safe to typecast C and C++ complex types because they're explicitly speced to have the same memory layout), but I suppose the fragility will stay. I appreciate the cmake approach, but beware that it appears to break sometimes.

I can add a flag that allows the end user application (in this case the python interface to GraphBLAS) to take control from cmake and bypass the #cmakedefine's in GraphBLAS/Config/GraphBLAS.h.in.  Something like this:

https://github.com/DrTimothyAldenDavis/SuiteSparse/commit/e81e6c1dc800e6843d344711bb73fd262267e3a2

If that works (I'm still testing it), I can add it to GraphBLAS 9.3.0.beta2.  Would that work for python-graphblas?

For the JIT, you'd need to add something after GrB_init, which would use GrB_get to get the current C flags, then GrB_set to append -DGXB_CMAKE_BYPASS -DGXB_HAVE_COMPLEX_C99 (for example).  A bit tedious of course but the end python user wouldn't see it.

**Comment by alugowski** [comment]
> If that works (I'm still testing it), I can add it to GraphBLAS 9.3.0.beta2. Would that work for python-graphblas?

My comment about complex values is for other GraphBLAS users, not us. We have the luxury of known build environments and already have a place where I have to do a bunch of patching for MSVC complex values. The workaround for us is already in this PR and works on 9.1: https://github.com/GraphBLAS/python-suitesparse-graphblas/blob/50364241f0f7f2489eca69937dc85566f25d4807/build_graphblas_cffi.py#L85

I pointed it out because GraphBLAS is already a high-friction library and that small problem is just another thing a new user has to solve. It seems like something that can be solved automatically with no intervention needed at all. Sounds like I may have missed some interaction between GNU and MS tools that you found, and in that case I defer to what you found to be necessary. However as you can see the upstream method didn't work for us and it required manual intervention.

**Comment by DrTimothyAldenDavis** [comment]
> I pointed it out because GraphBLAS is already a high-friction library and that small problem is just another thing a new user has to solve. It seems like something that can be solved automatically with no intervention needed at all. Sounds like I may have missed some interaction between GNU and MS tools that you found, and in that case I defer to what you found to be necessary. However as you can see the upstream method didn't work for us and it required manual intervention.

It can get complicated, with all the system variations we see in SuiteSparse; the github CI there tries to account for them and test them.  I realize that the complex variable type thing can be a painful friction.

What I have done so far is to make sure your workaround will always work.  If an end user package is compiled with GraphBLAS, and they want to insist to see a particular complex type on their side, they can do either

```
#define GxB_HAVE_COMPLEX_C99
#include "GraphBLAS.h"
```

or

```
#define GxB_HAVE_COMPLEX_MSVC
#include "GraphBLAS.h"
```

or, to add the corresponding -D(whatever) to the compile command.  This is only needed if the configuration of GraphBLAS by cmake differs from what the user wants in their application.

So I've added this to GraphBLAS.h:

https://github.com/DrTimothyAldenDavis/GraphBLAS/blob/c634199269d46bd63dba9583b5692850e77c5170/Config/GraphBLAS.h.in#L166-L191

**Review by eriknw — APPROVED** [review]
LGTM, thanks so much @alugowski! I made a few tweaks:
- Build with numpy 2
- Add Python 3.13
- Drop Python 3.8
- Update `cibuildwheel` b/c `yum install` stopped working due to centos mirrors being EOLed

Care to take a quick look before we merge? Also, should we jump straight to SuiteSparse:GraphBLAS 9.3.1 (the latest)?

**Comment by DrTimothyAldenDavis** [comment]
GraphBLAS 9.3.1 has 3 bug fixes for bugs in 9.1.0. So going to 9.3.1 would be a good idea.

**Comment by DrTimothyAldenDavis** [comment]
See these in the GraphBLAS/Doc/ChangeLog:

9.3.1:

  * (56) bug fix:  wrong type for fgetc return value in JITpackage; leads to infinite loop on some systems when building GraphBLAS.

Aug 2, 2024: version 9.3.0

    * (55) bug fix: GrB_apply with user-defined index-unary op would fail if A was iso and the JIT was disabled (failure in generic case). Caught by Christina Koutsou.
    * (54) bug fix: reducing a huge iso full matrix to a scalar resulted in integer overflow if nrows*ncols was larger than about 2^60.

File: .github/workflows/wheels.yml
Comment (alugowski) [inline .github/workflows/wheels.yml:96]: If we're dropping Python 3.8 then this is no longer needed

File: .github/workflows/wheels.yml
Comment (alugowski) [inline .github/workflows/wheels.yml:150]: If a point release is specified then dependabot will create update PRs for every new point release. We probably want to stick with `2.21` instead. That will pull in the latest point release.

File: pyproject.toml
Comment (alugowski) [inline pyproject.toml:70]: Is this compatible with numpy 2?

**Comment by alugowski** [comment]
> LGTM, thanks so much @alugowski! I made a few tweaks:
> 
>     * Build with numpy 2
> 
>     * Add Python 3.13
> 
>     * Drop Python 3.8
> 
>     * Update `cibuildwheel` b/c `yum install` stopped working due to centos mirrors being EOLed
> 
> 
> Care to take a quick look before we merge? Also, should we jump straight to SuiteSparse:GraphBLAS 9.3.1 (the latest)?

Looks great!

We should jump to latest before releasing, but in a separate PR.

File: pyproject.toml
Comment (eriknw) [reply pyproject.toml:70]: Yes. Wheels built with numpy 2 are compatible with older numpy versions, but the reverse is not true.

File: .github/workflows/wheels.yml
Comment (eriknw) [reply .github/workflows/wheels.yml:150]: Done, thanks

File: .github/workflows/wheels.yml
Comment (eriknw) [reply .github/workflows/wheels.yml:96]: Done

File: pyproject.toml
Comment (eriknw) [reply pyproject.toml:70]: See: https://numpy.org/doc/stable/dev/depending_on_numpy.html#numpy-2-0-specific-advice

**Comment by eriknw** [comment]
Thanks for the quick responses! The requested updates have been made. Let's merge after CI passes, then update to latest SuiteSparse:GraphBLAS in a new PR.

**Pull request merged**
